File: src/schema/template.ts
import type { TemplateId, SlotTag, PartId } from './ids'
import type { Transform } from './transform'

export type SlotDefinition = {
  readonly tag: SlotTag
  readonly name: string
  readonly anchor: Transform
  readonly defaultPartId: PartId | undefined
  readonly pairedSlot: SlotTag | undefined
  readonly required: boolean
}

export type Template = {
  readonly id: TemplateId
  readonly name: string
  readonly description: string
  readonly slots: readonly SlotDefinition[]
  readonly version: number
}

export const slotByTag = (
  template: Template,
  tag: SlotTag,
): SlotDefinition | undefined =>
  template.slots.find((slot) => slot.tag.value === tag.value)

File: src-tauri/src/schema/template.rs
use serde::{Deserialize, Serialize};
use ts_rs::TS;

use super::ids::{PartId, SlotTag, TemplateId};
use super::vec::Transform;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, TS)]
#[ts(export)]
pub struct SlotDefinition {
    pub tag: SlotTag,
    pub name: String,
    pub anchor: Transform,
    #[serde(rename = "defaultPartId")]
    pub default_part_id: Option<PartId>,
    #[serde(rename = "pairedSlot")]
    pub paired_slot: Option<SlotTag>,
    pub required: bool,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, TS)]
#[ts(export)]
pub struct Template {
    pub id: TemplateId,
    pub name: String,
    pub description: String,
    pub slots: Vec<SlotDefinition>,
    pub version: u32,
}
